lib-php/Hierarchy.php

```diff
@@ -0,0 +1,150 @@
+                    TO_CHAR(
+                        TO_TIMESTAMP(EXTRACT(epoch FROM p2.indexed_date)),
+                        'YYYY-MM-DD\"T\"HH:MI:SS+00:00'
+                    ) AS indexed_date
+                FROM placex AS p2
+                WHERE p2.country_code = :countryCode
+                AND p2.class = 'boundary'
+                AND p2.type = 'administrative'
+                AND p2.osm_type = 'R'
+                AND admin_level = (
+                    SELECT DISTINCT p1.admin_level
+                    FROM placex AS p1
+                    WHERE p1.country_code = :countryCode
+                    AND p1.class = 'boundary'
+                    AND p1.type = 'administrative'
+                    AND p1.osm_type = 'R'
+                    AND p1.admin_level > CASE
+                        WHEN p1.country_code = 'mt' THEN 3
+                        ELSE 2
+                        END
+                    ORDER BY p1.admin_level ASC
+                    LIMIT 1
+                )
+            ";
+
+            $aResult = $this->oDB->getAll($sSQL, [
+                ':countryCode' => $sCountryCode,
+            ]);
+        } else {
+            $sSQL = "
+                SELECT
+                    p2.osm_type,
+                    p2.osm_id,
+                    COALESCE(
+                        p2.name -> 'int_name',
+                        p2.name -> 'alt_name',
+                        p2.name -> 'name'
+                    ) AS name,
+                    TO_CHAR(
+                        TO_TIMESTAMP(EXTRACT(epoch FROM p2.indexed_date)),
+                        'YYYY-MM-DD\"T\"HH:MI:SS+00:00'
+                    ) AS indexed_date
+                FROM place_addressline AS pa
+                JOIN placex AS p1
+                    ON pa.address_place_id = p1.place_id
+                JOIN placex AS p2
+                    ON pa.place_id = p2.place_id
+                WHERE p1.osm_id = :osmId
+                AND p1.osm_type = 'R'
+                AND pa.isaddress
+                AND pa.fromarea
+                AND p2.admin_level = :adminLevel
+                AND p2.class = 'boundary'
+                AND p2.type = 'administrative'
+                AND p2.osm_type = 'R'
+            ";
+
+            $aResult = $this->oDB->getAll($sSQL, [
+                ':osmId' => $iOsmId,
+                ':adminLevel' => $iAdminLevel,
+            ]);
+        }
+
+        $aReturn = [];
+
+        if ($aResult) {
+            foreach ($aResult as $aResultItem) {
+                $aResultItem['children'] = $this->getChildrenRecursively(
+                    $aResultItem['osm_id'],
+                    $iAdminLevel + 1,
+                    $sCountryCode
+                );
+
+                $aReturn[] = $aResultItem;
+            }
+        } else {
+            $aReturn = $this->getChildrenRecursively(
+                $iOsmId,
+                $iAdminLevel + 1,
+                $sCountryCode
+            );
+        } 
```

Comment:
ich lese hier heraus, dass die admin-level immer lückenlos sein müssen, halte das für eine gefährliche Annahme...

bin mir nicht 100% sicher, wie das N/A genau zu verstehen ist, aber es ist denkbar, dass mittendrin was fehlt. Wir sollten womöglich spätestens beim Import eine große Änderung der Anzahl an Nodes in der Hierarchie blockieren (force flag oder sowas, um trotzdem zu pushen?)
![image](https://user-images.githubusercontent.com/70581149/135064713-2b6654c7-0b95-486d-be39-7dc6c4883543.png)